packages/http-server-javascript/src/common/reference.ts

```diff
@@ -0,0 +1,296 @@
+      if (isArrayModelType(ctx.program, type)) {
+        const argumentType = type.indexer.value;
+
+        const argTypeReference = emitTypeReference(ctx, argumentType, position, module, {
+          altName: options.altName && getArrayElementName(options.altName),
+        });
+
+        if (isImportableType(ctx, argumentType) && argumentType.namespace) {
+          module.imports.push({
+            binder: [argTypeReference],
+            from: createOrGetModuleForNamespace(ctx, argumentType.namespace),
+          });
+        }
+
+        return asArrayType(argTypeReference);
+      }
+
+      // Now other well-known models.
+      if (isWellKnownModel(ctx, type)) {
+        return emitWellKnownModel(ctx, type, module, options.altName);
+      }
+
+      // Try to reduce the model to an effective model if possible.
+      const effectiveModel = getEffectiveModelType(ctx.program, type);
+
+      if (effectiveModel.name === "") {
+        // We might have seen the model before and synthesized a declaration for it already.
+        if (ctx.syntheticNames.has(effectiveModel)) {
+          // TODO/witemple: I feel like I'm missing an import here.
+          const name = ctx.syntheticNames.get(effectiveModel)!;
+          module.imports.push({
+            binder: [name],
+            from: ctx.syntheticModule,
+          });
+          return name;
+        }
+
+        // Require preferredAlternativeName at this point, as we have an anonymous model that we have not visited.
+        if (!options.altName) {
+          return emitModelLiteral(ctx, effectiveModel, module);
+        }
+
+        // Anonymous model, synthesize a new model with the preferredName
+        ctx.synthetics.push({
+          kind: "anonymous",
+          name: options.altName,
+          underlying: effectiveModel,
+        });
+
+        module.imports.push({
+          binder: [options.altName],
+          from: ctx.syntheticModule,
+        });
+
+        ctx.syntheticNames.set(effectiveModel, options.altName);
+
+        return options.altName;
+      } else {
+        // The effective model is good for a declaration, so enqueue it.
+        ctx.typeQueue.add(effectiveModel);
+      }
+
+      const friendlyName = getFriendlyName(ctx.program, effectiveModel);
+
+      // The model may be a template instance, so we generate a name for it.
+      const templatedName = parseCase(
+        friendlyName
+          ? friendlyName
+          : effectiveModel.templateMapper
+            ? effectiveModel
+                .templateMapper!.args.map((a) => ("name" in a ? String(a.name) : ""))
+                .join("_") + effectiveModel.name
+            : effectiveModel.name
+      );
+
+      if (!effectiveModel.namespace) {
+        throw new Error("UNREACHABLE: no parent namespace of named model in emitTypeReference");
+      }
+
+      // TODO/witemple: I believe this is going to declare all template instances in the module of the template itself, which
```

Comment:
todo

Reply:
After reviewing I don't think this is a problem, or that there's a better solution to what we're doing at the moment. Might revisit in the future if it's a problem.